File: app/src/main/java/com/tunes/player/loaders/ItemsLoader.java
package com.tunes.player.loaders;

import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.tunes.player.activities.DetailsActivity;
import com.tunes.player.interfaces.AsyncTaskCallback;
import com.tunes.player.model.MusicModel;
import com.tunes.player.singleton.TrackManager;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/** Replaced deprecated AsyncTask with ExecutorService + main-thread Handler. */
public class ItemsLoader {

    private static final String TAG = "ItemsLoader";
    private static final ExecutorService sExecutor = Executors.newSingleThreadExecutor();
    private static final Handler sMainHandler = new Handler(Looper.getMainLooper());

    private final AsyncTaskCallback.Simple mCallback;
    private final String mTitle;
    private final int mChoice;

    public ItemsLoader(AsyncTaskCallback.Simple callback, String itemToSearch, int choice) {
        this.mCallback = callback;
        this.mTitle    = itemToSearch;
        this.mChoice   = choice;
    }

    public void execute() {
        sExecutor.execute(() -> {
            List<MusicModel> result = filter();
            sMainHandler.post(() -> {
                if (mCallback != null) mCallback.onTaskComplete(result);
            });
        });
    }

    private List<MusicModel> filter() {
        List<MusicModel> source = TrackManager.getInstance().getMainList();
        List<MusicModel> out = new ArrayList<>();
        if (source == null) return out;

        for (MusicModel md : source) {
            if (mChoice == DetailsActivity.CATEGORY_ALBUM) {
                if (md.getAlbum() != null && md.getAlbum().equals(mTitle)) out.add(md);
            } else if (mChoice == DetailsActivity.CATEGORY_ARTIST) {
                if (md.getArtist() != null && md.getArtist().equals(mTitle)) out.add(md);
            }
        }

        if (out.isEmpty()) Log.d(TAG, "No items found for: " + mTitle);
        return out;
    }
}

File: app/src/main/java/com/tunes/player/loaders/ItemsLoader.kt
package com.tunes.player.loaders

import android.util.Log
import com.tunes.player.model.MusicModel
import com.tunes.player.singleton.TrackManager
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

class ItemsLoader(private val title: String, private val category: Int) {

    companion object {
        const val CATEGORY_ALBUM = 1
        const val CATEGORY_ARTIST = 2
    }

    suspend fun loadItems(): List<MusicModel> = withContext(Dispatchers.IO) {
        val source = TrackManager.instance.getMainList()
        val out = mutableListOf<MusicModel>()

        for (md in source) {
            when (category) {
                CATEGORY_ALBUM -> if (md.album == title) out.add(md)
                CATEGORY_ARTIST -> if (md.artist == title) out.add(md)
            }
        }

        if (out.isEmpty()) Log.d("ItemsLoader", "No items found for: $title")
        out
    }
}
